File: leetcode/python/longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        left = 0
        char_freq: defaultdict[str,int] = defaultdict(int)
        for right in range(len(s)):
            char_freq[s[right]]+=1
            while char_freq[s[right]]>1:
                char_freq[s[left]]-=1
                left+=1
            max_length = max(max_length, (right-left+1))
        return max_length

File: leetcode/java/LongestSubstringWithoutRepeatingCharacters.java
import java.util.HashSet;
import java.util.Set;

public class LongestSubstringWithoutRepeatingCharacters {

    public int lengthOfLongestSubstring(String s){
        Set<Character> st = new HashSet<>();
        int n = s.length();
        int l = 0, r = 0;
        int maxLength = 0;
        while (r < n) {
            while(st.contains(s.charAt(r))){
                st.remove(s.charAt(l));
                l++;
            }
            st.add(s.charAt(r));
            maxLength = Integer.max(maxLength, r - l +1);
            r++;
        }
        return maxLength;
    }
}